[AndrewOng2066] iP

AndrewOng2066 opened the pull request:
# Andelu Task Bot
> [!NOTE]
> Life lacks excitement without tasks - Andelu

Andelu Bot is here to help you keep track of all your tasks, ensuring your life is both meaningful and seamless.
It is:

- Text based
- Easy to learn 
- ~~Fast~~ _SUPER SUPER FAST_ to use

## Getting Started 💻

To use  Andelu Bot, follow these steps:

1. Download it from [here](https://github.com/AndrewOng2066/ip/releases/tag/A-Jar) (`jar` File)
2. Run the `jar` file (double click it) or Run it from the command line with `java -jar IndividualProject.jar`
3. Add your tasks (Todo, Deadline, Event)
4. Let Andelu Bot manage your tasks😉

And it is **FREE** to use! 🚫💵

## Features

- [x] Managing ToDos
- [x] Managing deadlines 
- [x] Managing Events
- [x] Search Tasks
- [x] Search Tasks based on date
- [ ] GUI
- [ ] Editing tasks

## Java Developer Guide
Entry point for Andelu Bot (Duke.java):
```
public static void main(String[] args) {
        new Duke("duke").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Implement an initial skeletal version of it that simply greets the user and exits.

Commit:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

Commit:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye. Applied indentation

Commit:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

Commit:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

Commit:
Add support for tracking three types of tasks

Commit:
Use the input/output redirection technique to semi-automate the testing of Duke.
Commented the code for the "else" portion

Commit:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

Commit:
Add support for deleting tasks from the list.
Added exception handling for integer input.
replaced the code with the individual method.

Commit:
Added Enumerations for the function

Commit:
Added a test case but not working due to the spaces after ':'.
new files added: ACTUAL2.TXT, EXPECTED2.TXT, input2.txt, runtest2.bat

Commit:
Fixed the spaces issue, i.e. spaces after ":"
edited the EXPECTED2.TXT to remove the spaces.

Commit:
Fixed some changes to the EXPECTED.TXT due to the previously fixed bug (i.e. spaces)

Commit:
Added new exception handling (i.e. ensure that the input integer are more than 0)

Commit:
Added shell file (runtest.sh)

Commit:
Add a new function to store data into a file. However, the current llistMethod is not reading from the file. Add new files, i.e. duke.txt, DukeException.java, FileManager.java

Created a new DukeException.java and moved the DukeException code over.

Created a new FileManager to write data to a txt file.

Commit:
Change the layout of each object class (ToDo, Deadline, Event)

Commit:
Complete the save feature without JavaDoc

Commit:
Add Javadoc to the program

Commit:
Merge branch 'branch-Level-7'

Commit:
Add a new empty line

Commit:
Merge branch 'branch-Level-7'

Commit:
Add a new method to convert String to LocalDateTime

Commit:
Add a new method to convert LocalDateTime to String.

Commit:
Add a method that satisfied the minimal goal for level-8

Commit:
Add a method to satisfy the stretch goal for level-8

Commit:
Merge branch 'branch-Level-8'

Commit:
Change Code for "MoreOOP"
Major Change

Commit:
Merge branch 'branch-MoreOOP'

Commit:
Divide classes into packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Add Gradle file

Commit:
Add Junit tests

Commit:
Add META-INF/MANIFEST.MF

Commit:
Add JavaDoc comments to the code

Commit:
Update more Javadoc and Java coding style

Commit:
Change the typo

Commit:
Reverts some Java Coding Style (Should have done this at branch-A-CodingStandard)

Commit:
Update the code with the coding style

Commit:
Add a feature to let the user to find tasks based on the description

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/command/Command.java
#	src/main/java/command/OtherCommand.java
#	src/main/java/command/SearchCommand.java
#	src/main/java/duke/Ui.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java

Commit:
Resolve the typo

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix the missing import on Parser.java

Commit:
Fix the typo on FindCommand.java

Commit:
Remove extra line on OtherCommand.java

Commit:
Rename Duke.java to Andelu.java
Rename the package from duke to andelu

Commit:
Fix some bugs after changing the class name from Duke.java to Andelu.java

Commit:
Change the code to meet the checkstyle condition

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add a GUI to the chatbot using JavaFX

Commit:
Merge branch 'branch-Level-10'

Commit:
Edit the margin size for the label in DialogBox fxml

Joelwang22 commented on src/test/java/command/AddEventCommandTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Test
+    public void executeCommand_missingOneDateTime_exceptionThrown() {
+        DukeException exception = assertThrows(DukeException.class, () -> {new AddEventCommand("event meeting /from 2024-10-01 23:59").excuteCommand(new TaskList(), new Ui(), new Storage("duke"));});
```
I like how thorough you are with testing! However the lines where "exception" variable are is longer than 120 characters in length which is the java coding standards, I'd recommend putting in some line breaks, maybe before "() ->". Other than that it looks good!

Joelwang22 reviewed (commented):
Looks Good! 👍 
Adding line breaks would improve readability!

Joelwang22 commented on src/main/java/command/AddEventCommand.java:
```diff
@@ -1,16 +1,17 @@
-import duke.TaskList;
-import duke.Ui;
-import duke.DateTimeManager;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
 
+import andelu.AndeluException;
+import andelu.DateTimeManager;
+import andelu.Storage;
+import andelu.TaskList;
+import andelu.Ui;
 import task.Event;
 import task.Task;
```
Good job listing imported classes explicitly!

Joelwang22 reviewed (commented):
Code is well written with no obvious coding standard violations! 💯

yyccbb commented on src/main/java/andelu/Andelu.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return Ui.
+     */
+    public Ui getUi() {
+        return ui;
+    }
```
### #Class structure
Maybe you could consider using singleton pattern to save the trouble?

yyccbb commented on src/main/java/andelu/Andelu.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return Storage.
+     */
+    public Storage getStorage() {
+        return storage;
+    }
```
### #Class structure
Same as above.

yyccbb commented on src/main/java/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,75 @@
+        if (splitInput.length <= 1) {
+            throw new AndeluException("Missing the description!");
+        }
+        String[] deadlineSplit = input.split("/");
+        if (deadlineSplit.length < 2) {
+            throw new AndeluException("Invalid format for new Deadline!");
+        }
+        String name = deadlineSplit[0].substring(9).trim();
+        String by = deadlineSplit[1].substring(3).trim();
```
### #General
Interesting way of parsing! I may consider using this in my project as well. Thanks for the inspiration!

yyccbb commented on src/main/java/andelu/Storage.java:
```diff
@@ -0,0 +1,211 @@
+    /**
+     * Creates directory and file if it doesn't exist.
+     */
+    private void createFile() {
+        try {
+            if (!Files.exists(DIRECTORY_PATH)) {
+                Files.createDirectories(DIRECTORY_PATH);
+            }
```
### #General
Good job identifying the need to create the directory when it is absent. I have missed this in my code.

yyccbb commented on src/main/java/andelu/Storage.java:
```diff
@@ -0,0 +1,211 @@
+     * @param tasks An arrayList of Tasks.
+     * @return String format of the list of Tasks.
+     */
+    private String convertTasksToString(ArrayList<Task> tasks) {
+        StringBuilder result = new StringBuilder();
+
+        for (Task task: tasks) {
+            if (task instanceof ToDo) {
+                result.append("T | ");
+                result.append(task.getStatusIcon().equals("X") ? "1 | " : "0 | ");
+                result.append(task.getDescription());
+                result.append(System.getProperty("line.separator"));
+            } else if (task instanceof Deadline) {
+                result.append("D | ");
+                result.append(task.getStatusIcon().equals("X") ? "1 | " : "0 | ");
+                result.append(task.getDescription() + " | ");
+                result.append(((Deadline) task).getBy());
+                result.append(System.getProperty("line.separator"));
+            } else {
+                result.append("E | ");
+                result.append(task.getStatusIcon().equals("X") ? "1 | " : "0 | ");
+                result.append(task.getDescription() + " | ");
+                result.append(((Event) task).getStart() + " | ");
+                result.append(((Event) task).getEnd());
+                result.append(System.getProperty("line.separator"));
+            }
+
+        }
+
+        return result.toString();
+    }
```
### #General
I really like how you format your code. It is very clean and readable!

yyccbb commented on src/main/java/andelu/Andelu.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    /*
+    /**
+     * Runs the program.
+     *
+    public void run() {
+        ui.printWelcome();
+        boolean isExit = false;
+
+        while (!isExit) {
+            try {
+                String command = ui.readCommand();
+                ui.printOpeningDottedLine();
+
+                Command c = Parser.parse(command);
+                c.executeCommand(tasks, ui, storage);
+                isExit = c.getIsExit();
+            } catch (AndeluException e) {
+                ui.printError(e.getMessage());
+            } finally {
+                ui.printClosingDottedLine();
+            }
+        }
+    }
+    */
```
### #Comment
I know this is temporary but just remember to remove them later:)

yyccbb reviewed (commented):
LGTM! I really like how neat your codebase is. Keep up the good work 💯

Commit:
Update build.gradle with new directory for application segment

The file, build.gradle, is outdated due to the newly implementation of
javaFx.

This change is required when creating a jar file.

The argument of mainClass is changed from "andelu.Andelu" to
"gui.Launcher".

Since the main entry of the program is changed due to javaFx, we
need to update the new main entry.

Commit:
Add assert to the program

There is no assertion in the program to document important
assumptions.

This assertion helps to identify any potential error caused by
programmer.

Assertion has been added to each subclass of command.java and
storage.java.

The assumption made to these classes helps to ensure that the values
are correct or not null.

Commit:
Improve code quality

Some codes do not fulfil the criteria for code quality.

Took out a repeated variable in multiple methods and treated it as
field.

Since these codes are the same, we can assign them as a field.

Commit:
Change the code in executeCommand of ListCommend to use Java Stream

AndrewOng2066 committed:
Merge pull request #1 from AndrewOng2066/branch-A-Streams

Change the code in executeCommand of ListCommend to use Java Stream

Commit:
Edit the duke.txt due to some testings

AndrewOng2066 committed:
Merge branch 'master' into branch-A-CodeQuality

AndrewOng2066 committed:
Merge pull request #2 from AndrewOng2066/branch-A-CodeQuality

Improve code quality

AndrewOng2066 committed:
Merge branch 'master' into branch-A-Assertions

AndrewOng2066 committed:
Merge pull request #3 from AndrewOng2066/branch-A-Assertions

Add assert to the program

Commit:
Add gradle.yml in .github/workflows directory

Commit:
Add a new feature to set priority to each task

Current program does not allow user to attach a priority level to
each task.

This change allow the user to indicate which tasks are high, medium,
low or default.

Each subclass of Task has a new field called "priorityLevel".
AddCommand class has edited to attach the priority level.

Each subclass of Task should have the field, "priorityLevel", hence
we need to edit the code on these classes.

Commit:
Add a new feature to set priority to each task

Current program does not allow user to attach a priority level to
each task.

This change allow the user to indicate which tasks are high, medium,
low or default.

Each subclass of Task has a new field called "priorityLevel".
AddCommand class has edited to attach the priority level.

Each subclass of Task should have the field, "priorityLevel", hence
we need to edit the code on these classes.

Commit:
Merge branch 'branch-C-Priority'

Commit:
Fix a bug that causes an issue when reading from the txt file

Commit:
Improve the GUI design

Commit:
Improve the GUI design

Commit:
Merge branch 'branch-A-BetterGUI'

Commit:
Update the application name and GUI

Commit:
Update the application name and GUI

Commit:
Update the txt filename to "andeluFile"

Commit:
Change the background of the application

Commit:
Add Ui.png under docs directory

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

Commit:
Add new data to the txt file

Commit:
Merge branch 'master' of https://github.com/AndrewOng2066/ip

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

Commit:
Change the exit feature to close the program

AndrewOng2066 committed:
Update README.md

Commit:
Remove all CLI print statement

Commit:
Change Jar filename

Commit:
Edit the error message for adding tasks

Commit:
Remove all unnecessary code

Commit:
Change the exit behavior

Commit:
Update the current Junit and add a new file for junit testing on
AddToDoCommand

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

Commit:
Update README.md

Commit:
Update andeluFile.txt

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md

AndrewOng2066 committed:
Update README.md
